url11fb0973cf_raichuman9_Swimmer526_tgz-pJ8-fr_inria_optimization_cmaes_examples_CMAExample1J8/src/ubc/swim/world/characters/RefTrajHumanChar.java

```diff
@@ -0,0 +1,541 @@
+	protected static final int NUM_PARAMS_PER_SHOULDER_TRAJECTORY = NUM_PARAMS_PER_SHOULDER_BASIS_FUNC * NUM_BASIS_FUNCS_PER_SHOULDER_TRAJECTORY;
+	
+	protected static final float MIN_STROKE_PERIOD = 0.2f; 
+	
+	//if true, the normalized angle of character's shoulder is used as phase input to other joint trajectories
+	//NOTE: if this changes, character must be re-optimized
+	protected static final boolean USE_RIGHT_SHOULDER_ANGLE_AS_PHASE = true;  
+	
+	//Body params
+	protected float height = 2; //2 meters
+	protected float headHeight = height / 8;
+	protected float headWidth = headHeight * 0.85f;
+	protected float armLen = height * 0.45f;
+	protected float upperArmLen = armLen * (1 / 2.2f);
+	protected float upperArmWidth = height / 16;
+	protected float lowerArmLen = armLen - upperArmLen;
+	protected float lowerArmWidth = upperArmWidth * 0.85f;
+	protected float legLen = height * 0.54f;
+	protected float upperLegLen = legLen * 0.46f;
+	protected float upperLegWidth = height / 10;
+	protected float lowerLegLen = legLen - upperLegLen;
+	protected float lowerLegWidth = upperLegWidth * 0.7f;
+	protected float footLen = height * 0.035f;
+	
+	protected Stroke stroke;
+	
+	protected float shoulderPeriod; //time period over which shoulders go through full rotation
+	
+	protected ArrayList<Body> rightBodies;
+	protected ArrayList<Body> leftBodies;
+	
+	protected ArrayList<Joint> joints;
+	protected ArrayList<Joint> leftJoints;
+	protected ArrayList<Joint> rightJoints;
+	
+	protected RevoluteJoint rightShoulderJoint = null; //angle of this joint is optionally used to drive phase
+	
+	protected float prevTorque = 0.0f;
+	
+	protected float targetRightShoulderPhase = 0.0f; //phase goal for right shoulder after last step()
+	
+	protected ArrayList<RefTrajectory> trajectories;
+	protected ArrayList<SineTrajectory> sineTrajectories;
+	protected ArrayList<PolynomialTrajectory> shoulderTrajectories;
+	
+	/**
+	 * Create a new human character with given stroke
+	 * shoulderPeriod specifies the time interval over which shoulder joints try to 
+	 * execute a full rotation.
+	 * @param stroke
+	 */
+	public RefTrajHumanChar(Stroke stroke, float shoulderPeriod) {
+		super();
+		
+		this.stroke = stroke;
+		this.shoulderPeriod = shoulderPeriod;
+	
+		rightBodies = new ArrayList<Body>();
+		leftBodies = new ArrayList<Body>();
+		
+		joints = new ArrayList<Joint>();
+		
+		leftJoints = new ArrayList<Joint>();
+		rightJoints = new ArrayList<Joint>();
+		
+		trajectories = new ArrayList<RefTrajectory>();
+		sineTrajectories = new ArrayList<SineTrajectory>();
+		shoulderTrajectories = new ArrayList<PolynomialTrajectory>();
+	}
+
+	@Override
+	public int getNumControlDimensions() {
+		//NOTE: shoulder trajectory is manually coded and not included in control dims
+		
+		int numJoints = 1; //1 for head
+		numJoints += 1; //elbow 
+		numJoints += 2; //hip & knee
+		return numJoints * NUM_PARAMS_PER_SINE_TRAJECTORY;
+	}
+	
+	@Override
+	public void initialize(World world) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 126 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 345-350 into a separate method.